xds/src/main/java/io/grpc/xds/internal/rbac/engine/AuthorizationEngine.java

```diff
@@ -0,0 +1,208 @@
+import io.grpc.xds.internal.rbac.engine.cel.RuntimeTypeProvider;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * CEL Evaluation Engine is part of the authorization framework in gRPC.
+ * CEL Evaluation Engine takes one or two Envoy RBAC policies as input
+ * and uses CEL library to evaluate the condition field 
+ * inside each RBAC policy based on the provided Envoy Attributes. 
+ * CEL Evaluation Engine will generate an authorization decision which
+ * could be ALLOW, DENY or UNKNOWN.
+ * 
+ * <p>Use as in:
+ * 
+ * <pre>
+ *  AuthorizationEngine engine = new AuthorizationEngine(
+ *      ImmutableList.of(rbacPolicies));
+ *  AuthorizationDecision result = engine.evaluate(new EvaluateArgs(call, headers));
+ * </pre>
+ */
+public class AuthorizationEngine {
+  private static final Logger log = Logger.getLogger(AuthorizationEngine.class.getName());
+  
+  /**
+   * RbacEngine is an inner class that holds RBAC action 
+   * and conditions of RBAC policy.
+   */
+  private static class RbacEngine {
+    @SuppressWarnings("UnusedVariable")
+    private final Action action;
+    private final ImmutableMap<String, Expr> conditions;
+
+    public RbacEngine(Action action, ImmutableMap<String, Expr> conditions) {
+      this.action = Preconditions.checkNotNull(action);
+      this.conditions = Preconditions.checkNotNull(conditions);
+    }
+  }
+
+  private final RbacEngine allowEngine;
+  private final RbacEngine denyEngine;
+
+  /**
+   * Creates a CEL-based Authorization Engine from one Envoy RBAC.
+   * @param rbacPolicy input Envoy RBAC policy.
+   */
+  public AuthorizationEngine(RBAC rbacPolicy) {
+    Map<String, Expr> conditions = new LinkedHashMap<>();
+    for (Map.Entry<String, Policy> policy: rbacPolicy.getPolicies().entrySet()) {
+      conditions.put(policy.getKey(), policy.getValue().getCondition());
+    }
+    allowEngine = (rbacPolicy.getAction() == Action.ALLOW) 
+        ? new RbacEngine(Action.ALLOW, ImmutableMap.copyOf(conditions)) : null; 
+    denyEngine = (rbacPolicy.getAction() == Action.DENY) 
+        ? new RbacEngine(Action.DENY, ImmutableMap.copyOf(conditions)) : null; 
+  }
+
+  /**
+   * Creates a CEL-based Authorization Engine from two Envoy RBACs.
+   * When it takes two RBACs, the order has to be a RBAC with DENY action
+   * followed by a RBAC with ALLOW action.
+   * @param denyPolicy input Envoy RBAC policy with DENY action.
+   * @param allowPolicy input Envoy RBAC policy with ALLOW action.
+   * @throws IllegalArgumentException if the user inputs an invalid RBAC list.
+   */
+  public AuthorizationEngine(RBAC denyPolicy, RBAC allowPolicy) throws IllegalArgumentException {
+    if (denyPolicy.getAction() != Action.DENY || allowPolicy.getAction() != Action.ALLOW) {
+      throw new IllegalArgumentException( "Invalid RBAC list, " 
```

Comment:
Consider using `com.google.common.base.Preconditions.checkArgument` like it is used everywhere else in grpc-java